README.txt
OpenCAN v0.8 - Open software platform for controlling CAN devices

Welcome to the OpenCAN community!

== What is OpenCAN? ==
OpenCAN is an open source platform for building Windows, Mac and Linux software that can interact with various Controller Area Network (CAN) buses and devices. It provides an abstract C++ interface that can be used to control the basic functions of any CAN device. Support for specific devices are written as OpenCAN plugins and loaded through a simple API call. So the hardware CAN drivers can be accessed through a generic interface, and application code can request a specific driver or use any that are available. Each component is cross-platform, enabling the efficient development of CAN software on Windows, Mac OS X and Linux. Everyone is free to use and extend OpenCAN according to the GNU Lesser General Public License (LGPL).


== Benefits ==
Simplifies adding CAN communication to software
Write one set of code that supports multiple devices and OSes
Add, remove or change CAN devices/drivers without modifying application code or recompiling
Simulate CAN devices using a simple C++ API instead of proprietary code
Reduces development and testing time by providing a simulated CAN bus to develop and test application code
Quickly prototype, test and demo software without device hardware
Reduces workspace clutter - no wires needed for simulation
Enables a greater separation between networking and application code
Minimizes hardware-specific code to a few simple methods
Reduces maintenance time and cost
Enables backward and forward compatibility with CAN devices, CAN driver APIs, OSes and hardware



== Features ==
Supports Windows, Mac OS X and Linux
Light-weight and simple API encapsulation of hardware CAN drivers
Fully documented API using Doxygen
Specific CAN hardware support can be statically compiled or loaded at runtime through plugins
Includes plugins for PEAK-System USB-PCAN devices on Windows and Linux as well as for Linux LinCAN devices
Includes the SimCAN device plugin for simulating a CAN bus across multiple applications
Includes a cross-platform GUI CAN bus analyzer and simulator called Testbench


== SimCAN Device Plugin ==
The cross-platform open source SimCAN device plugin uses an Inter-Process Communication (IPC) shared memory buffer to simulate a physical CAN bus that multiple applications can use to send and receive CAN messages between each other as if they were on a physical CAN bus. It can be used to develop and test application code before the hardware is ready or for fairly light-weight and simple communication between desktop applications. When the hardware arrives, simply replace the SimCAN plugin with a plugin for the physical CAN hardware. Since OpenCAN provides a generic interface to hardware CAN drivers, the applications will communicate over the physical CAN bus unmodified.

Applications using the SimCAN device plugin can be started and begin using the device in any order. The first application to start the simulated device will setup the IPC shared memory buffer, and other applications will simply begin using the buffer when they start the device. Similarly, the simulated CAN bus is not affected by the timing or order of an application stopping the device, and the shared memory buffer is not destroyed until all applications using it have stopped or closed.

The simulated CAN bus is not part of the core OpenCAN library. Every application that intends on communicating over the simulated CAN bus must use the OpenCAN library and the SimCAN device plugin. It is not impossible to use the IPC shared memory buffer without SimCAN, but to ensure interoperability and proper function across software developed by different people, it is strongly recommended against.

While the SimCAN device plugin simulates a physical CAN bus from the perspective of the application code that interacts with it, SimCAN does not try to simulate real-world conditions such as noise, physical limitations on transfer speed, power failures, etc. So an application that runs properly using the SimCAN plugin could fail to perform as expected when connected to a physical CAN bus, if the same ideal conditions are expected for the physical network.

The SimCAN device plugin has been used to simulate an entire network of up to 16 applications each sending a CAN message about once every second. The upper bound on devices and bandwidth is system-dependent, and the upper bounds are likely much higher than the those for a physical CAN bus.

The SimCAN device plugin is intended for desktop software use (Windows, Mac and Linux) and not for use on systems with contrained resources, but you are free to try if you want to.


== Testbench CAN Bus Analyzer and Simulator ==
Testbench is a cross-platform open source CAN bus analyzer and simulator for one or more networks. Testbench provides a test and simulation framework for software that incorporates CAN communication and serves as a source code example of using OpenCAN in an application. It can be used to test CAN drivers in development or interact with other applications using the SimCAN device plugin. When using the SimCAN device plugin, Testbench can be used to monitor and send CAN messages to other applications as if they were all on a network connected by a physical CAN bus. So Testbench can be a useful tool from the start of development through the entire life of the software.

Testbench is not required to use the SimCAN device plugin. Any two or more applications using the plugin can communicate via the simulated CAN bus.


== Requirements ==
Qt 4.6+/5.0+ (http://www.qt.io)
   The OpenCAN library and device plugins only require Qt's QtCore component
   Testbench requires the additional QtWidgets component for Qt5 or the QtGui component for Qt4

To compile and run the PCAN or LinCAN device plugins, the device driver and its library need to be installed first.

The PCAN device plugin requires the PCAN-Light API
    Windows: http://www.peak-system.com/Software-APIs.305.0.html?&L=1
        PCAN-USB package including the device driver and the PCAN-Basic API: http://www.peak-system.com/PCAN-USB.199.0.html?&L=1
        Create folder C:\PEAK
        Place the extracted folders and files Drivers, PCAN-Basic API, etc. into C:\PEAK
        Install the PCAN-USB driver according to the accompanying installation instructions
        Place the PCAN-Light API folder in C:\PEAK
        Copy the PCAN_USB.dll to the Testbench directory or add to library path
    Linux: http://www.peak-system.com/fileadmin/media/linux/index.htm
        Compile and install kernel driver and PCAN library according to the accompanying installation instructions


== Operating System and Compiler Support ==
Microsoft Windows 7 (32-bit) with Visual Studio 2010 SP1 or MinGW gcc 4.8.0
Microsoft Windows 8.1 (32-bit and 64-bit) with Visual Studio 2012 SP2
Apple Mac OS X 10.9 "Mavericks", Cocoa (64-bit) with the Apple-provided Clang
Linux kernel 3.0+ with GCC 4.6+ and X.org 7.6+

For Windows, at least Visual Studio 2010 SP 1 or MinGW gcc 4.8.0 must be used for C++11 support.

These are a subset of the Qt Reference Configurations for the latest Qt version at the time this version of OpenCAN is released (Qt 5.3). While OpenCAN should compile and run properly on any platform that a supported version of Qt does, the above OS and compiler combinations are the only combinations verified to run properly. See the Qt documentation for more information: http://qt-project.org/doc/qt-5/supported-platforms.html.

The Linux distributions that are used to test OpenCAN are Red Hat Enterprise Linux 6 and Fedora, but any Linux installation with the same versions of kernel, GCC and X.org or newer than what is provided in Ubuntu 11.10 should compile and run properly. See the Qt documentation for more information: http://qt-project.org/doc/qt-5/linux.html.


== Conditions of Use ==
The OpenCAN library as well as the LinCAN and PCAN device plugins are distributed under the GNU Lesser General Public License (LGPL) version 2.1 and may be used under version 2.1 of the LGPL or any later version.

Testbench and the SimCAN device plugin are distributed under the GNU General Public License (GPL) version 3.0 and may be used under version 3.0 of the GPL or any later version.

Commercial licenses can be provided for any or all of the software distributed in this OpenCAN package. 

The OpenCAN library, the device plugins and Testbench are provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.


== Build ==
Projects are built using Qt QMake and a C++ compiler.  Qt Creator is the best IDE for building OpenCAN, Testench and all of the plugins.

Specific build instructions can be found at docs/index.html


== Install ==
Currently, there is no installation method. Simply copy the compiled files to the desired location or leave them where they are. See the OpenCAN documentation for further details on installation and use.


Using OpenCAN in your projects? We'd like to hear from you. Please tell us what you do and don't like as well as what you'd like to see in the future by sending an email to feedback@opencan.org. We appreciate any help in making OpenCAN more useful to the community.

Thank you for using OpenCAN!



Copyright (C) 2005-2020 Code of Intelligence, LLC.
(http://www.codeofintelligence.com)
Contact: contact@opencan.org


CHANGELOG.txt
xxxxxxxx v0.8  Added SimCAN IPC device plugin for simulating a CAN bus

                                Put doxygen documentation on the website
                                Consider separating the distribution of Testbench from OpenCAN to clarify and enhance marketing for both and increase OpenCAN "footprint" (e.g., two webpages, one for OpenCAN and one for Testbench)
                                Clean up SimCAN code, TODOs, etc.
                                Make sure OPENCAN_EXPORT works on Windows for Qt 4 and 5 - was commented out in occore.h
                                Figure out if there's a lag in receive window when turning off ECU while still listening on Windows

                                Filters with different signals emitted by OcDevice or a signal with a varying parameter
                                Implement message filters for an OcDevice that implement the UI for TestBench's filters and possibly others like PCAN-Basic

                                Save received CAN messages to file using a data logging backend, so a window doesn't need to be displayed
                                Load saved messages and playback, or allow messages in a window to be played back (PCAN-Trace)
                                Add ability to send CAN messages at user-specified time intervals
                                Add ability to change the number of milliseconds to check for received messages in autonomous mode?
                                Add SimLink feature to link a SimCAN bus to a hardware bus, etc. PCAN-Basic

                                Add PingCAN BusTest that sends a message once a second unprompted and echos any message sent on the bus
                                    Do this as a very simple example of using the SimCAN driver and OpenCAN in general
                                    Should BusTest use a specific message ID to filter messages? Probably
                                    Could have it turn LEDs on and off or something
                                Create a command line executable in addition to or instead of BusTest
                                    Send a message on can bus using first device (default), device index or name
                                    Log received messages to a file using first device (default), device index or name
                                        Can we know if other OpenCAN programs, like say another instance of the cmd exec, send a message?
                                    Simple example of using OpenCAN instead of Testbench which will get more complex

                                From user email: "It would be nice if OpenCAN (testbench) would have a more sophisticated GUI. I would like to have:
                                - message filters [already supported]
                                - J1939 support
                                - export mode
                                - import mode
                                - CANopen support (eds file parsing, node send/recieve/status)"



                                +Need to store whether message is extended or not in OcSimCanDevice
                                +Add SocketCAN device plugin - this is the standard Linux CAN driver since kernel 2.6.25 (https://gitorious.org/linux-can) - needs libsocketcan
                                +Add Linux support to PCAN-Basic plugin
                                +Add an OpenCAN namespace? only if Qt 5 uses one
                                +Add PCAN-Basic device plugin
                                +Add ability to set baud rate to arbitrary bits per second in Testbench if supported by driver
                                +Look for ocplugins in application path as well as OPENCAN_DIR, so apps don't need to run from OpenCAN directory
                                +Reorganize at least the OpenCAN library code into a src directory so the source files don't overshadow other files/folders
                                +Figure out if baud rate should be a dropdown, is there a finite number of rates or any up to 1 Mbit/sec?
                                +Make sure the autonomous is easy to understand and efficient
                                +Double-check OcDevice::timerEvent() for how timestamp rollover is handled
                                +Move the entire timestamp calculation out to Testbench as it should never have been in the driver!
                                +Replace slots with Q_SLOTS and signals with Q_SIGNALS in public headers (http://qt-project.org/doc/qt-5/signalsandslots.html)
                                +Finish getting rid of copy for loops and replacing with std::copy
                                +Make sure menu and everything else still look correct on Linux/Windows
                                +Need to check if everything compiles on Qt 4 without change
                                +Replaced ReceiveWindow's QTextEdit with QPlainTextEdit and set maximumBlockCount to improve efficiency
                                +Wrap string literals with QStringLiteral
                                +Modified the OcDevice::getMessage method to clarify and simplify memory management
                                +Use monospace font for CAN receive window
                                +Ported to Qt 5
                                +API changes:
                                    OcMessage now stores the timestamp as a QTime instead of a quint32 to simplify the interpretation and comparison of timestamps
                                    OcDevice::getMessage() and the messageReceived() signal now use a reference to an OcMessage instead of a pointer to clarify and simplify memory management
                                    OcDevice refactored to simplify writing new plugins and using OcDevices
                                        Removed receivedMessage(), simply check getMessage() return value

include/ocdoc.h
/******************************************************************************
**
**  Copyright (C) 2005-2020 Code of Intelligence, LLC.
**  Contact: contact@opencan.org
**
**  GNU Lesser General Public License Usage
**  This file may be used under the terms of the GNU Lesser General Public
**  License version 2.1 as published by the Free Software Foundation and
**  appearing in the file LICENSE included in the packaging of this file.
**  Please review the following information to ensure the GNU Lesser General
**  Public License version 2.1 requirements will be met:
**  http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html.
**
**  Commercial License Usage
**  Licensees holding valid commercial OpenCAN licenses may use this file in
**  accordance with the commercial license agreement or other written agreement
**  between you and Code of Intelligence that governs your use of the Software.
**  For further information, use the contact form at
**  http://codeofintelligence.com/contact.
**
******************************************************************************/

#ifndef OCDOC_H
#define OCDOC_H

////////////////////////////////////////////////////////////////////////////////
//
// This file contains the first page of the OpenCAN API Doxygen documentation
//
////////////////////////////////////////////////////////////////////////////////

/**		\mainpage

  \section Introduction

  The following is documentation for compiling, installing and using the OpenCAN API.

  \section Compiling

  OpenCAN is distributed as five separate projects, each with their own QMake \a .pro file.
  <table>
  <tr><td>OpenCAN.pro</td><td>OpenCAN library</tr>
  <tr><td>ocplugins/SimCAN.pro</td><td>SimCAN device plugin <em>(All platforms)</em></tr>
  <tr><td>ocplugins/PCAN.pro</td><td>PCAN device plugin <em>(Windows and Linux)</em></tr>
  <tr><td>ocplugins/LinCAN.pro</td><td>LinCAN device plugin <em>(Linux only)</em></tr>
  <tr><td>Testbench/Testbench.pro</td><td>Testbench application</tr>
  </table>
  Only the OpenCAN library is compiled using the OpenCAN.pro project file, and it must be
  compiled first.
  The other projects are optional, but you will need an OpenCAN device plugin to use the
  CAN bus or the SimCAN device plugin for the simulated CAN bus.

  OpenCAN has been tested on the following configurations:
  <ul>
    <li>Microsoft Windows 7 (32-bit) with Visual Studio 2010 SP1 or MinGW gcc 4.8.0</li>
    <li>Microsoft Windows 8.1 (32-bit and 64-bit) with Visual Studio 2012 SP2</li>
    <li>Apple Mac OS X 10.9 "Mavericks", Cocoa (64-bit) with the Apple-provided Clang</li>
    <li>Linux kernel 3.0+ with GCC 4.6+ and X.org 7.6+</li>
  </ul>

  To compile and run the PCAN or LinCAN device plugin, the device driver and its library
  need to be installed first.

  Qt Creator is the recommended tool to compile for users new to the Qt library. Qt Creator
  can be used to compile and debug each project by opening the project's .pro file, or you
  can follow the platform-specific compilation instructions using QMake. Qt Creator invokes
  QMake for you and is often the easiest way to get started with OpenCAN.
  - \subpage compilewin
  - \subpage compilelin
  - \subpage compilemac

  \section Installing

  Currently, there is no installation method. Simply copy the compiled files to the
  desired location or leave them where they are. However, adding the OpenCAN include
  and library directories to the system include and library paths can simplify
  compilation of plugins and software using the OpenCAN library.

  Compiling the PCAN device plugin requires the latest version of the PCAN-Light API
  from PEAK-System. Unless the PCAN-Light include and library directories are added
  to the

  \section Using

  To use the OpenCAN API in another project, you must add the OpenCAN include and
  library directories to your system include and library paths or set the flags
  appropriately in the project's QMake file. The Testbench QMake .pro file may be a
  helpful example of how to set the directories.

  To use OpenCAN Testbench simply run the executable. Testbench does very little
  without an OpenCAN device plugin installed and connected to a device.

  \section Troubleshooting

  \subsection Loading Plugins

  If you are having trouble loading the PCAN plugin in Windows, you may need to
  copy the PCAN_USB.dll from the PEAK software into the OpenCAN folder with
  testbench.exe.

  Other reasons for a plugin not loading is if the plugin is built with a different
  version of Qt than the Qt used to build OpenCAN, or if the plugin is not built with the
  installed version of OpenCAN loaded at runtime.
**/

/** \page compilewin Compile for Windows
  \section vs Compile using Visual Studio
  You must have the Qt library built for a supported version of Visual Studio
  installed before compiling and QTDIR\bin must be in your PATH. This is done
  for you if you use the Qt installer.
  - Open a Visual Studio Command Prompt (64-bit if compiling for 64-bit)
  - Execute \a qmake -tp vc to create the Visual Studio project
  - Use Visual Studio to compile and debug the project

  \section MinGW Compile using mingw
  You must have the Qt library built for MinGW installed before compiling.
  - Open a Command Prompt
  - Change to the directory containing the QMake file for the project to compile.
  - Execute \a qmake to create the gmake Makefile
  - Execute \a mingw32-make
**/

/** \page compilelin Compile for Linux
  - Change to the directory containing the QMake file for the project to compile.
  - Execute \a qmake to create the gmake Makefile.
  - Execute \a make to build the project
**/

/** \page compilemac Compile for Mac OS X
  - Change to the directory containing the QMake file for the project to compile.
  - Execute \a qmake to create the XCode project.
  - Open the created XCode project to compile and debug
**/

#endif // OCDOC_H
